packages/tpetra/core/src/Tpetra_Experimental_BlockCrsMatrix_def.hpp

```diff
@@ -2135,56 +2184,71 @@ class GetLocalDiagCopy {
   template<class Scalar, class LO, class GO, class Node>
   void
   BlockCrsMatrix<Scalar, LO, GO, Node>::
-  copyAndPermute (const ::Tpetra::SrcDistObject& source,
-                  size_t numSameIDs,
-                  const Teuchos::ArrayView<const LO>& permuteToLIDs,
-                  const Teuchos::ArrayView<const LO>& permuteFromLIDs)
+  copyAndPermuteNew (const ::Tpetra::SrcDistObject& source,
+                     size_t numSameIDs,
+                     const Kokkos::DualView<const local_ordinal_type*, device_type>& permuteToLIDs,
+                     const Kokkos::DualView<const local_ordinal_type*, device_type>& permuteFromLIDs)
   {
-    using std::endl;
+    using ::Tpetra::Details::Behavior;
+    using ::Tpetra::Details::dualViewStatusToString; 
+    using ::Tpetra::Details::ProfilingRegion;
+
     typedef BlockCrsMatrix<Scalar, LO, GO, Node> this_type;
-    const bool debug = false;
+    
+    ProfilingRegion profile_region("Tpetra::BlockCrsMatrix::copyAndPermuteNew");  
 
-    if (debug) {
+    const bool debug = Behavior::debug();
+    const bool verbose = Behavior::verbose();
+
+    // Define this function prefix 
+    std::string prefix;
```

Comment:
I just want to use the prefix for regular error checking which is not guarded by debug or verbose mode. Some error checking must be placed and it can use the prefix to show which rank has the problem. I also saw the usage of unique_ptr. Is there any benefit in using unique_ptr wrapping ? Most Tpetra methods are coarse grain operations and it probably won't get a performance problem from using string.

Reply:
We've seen some performance issues with indiscriminate creation and copying of std::string to create timer labels.  Alas, MueLu is full of that so it's probably too late.  Whatever.

Reply:
Yeah.... if we abuse strings, there should be some side effects. In this code, most cout of prefix is only made in debug mode and I think it is okay.